runtime/lua/vim/_pack.lua

```diff
@@ -0,0 +1,631 @@
+--- @field dir string
+--- @field status vim._pack.Status
+--- @field hash string
+--- @field url string
+--- @field pin boolean?
+--- @field branch string?
+--- @field build string? | function?
+
+--- @nodoc
+--- @enum vim._pack.Status
+M.status = {
+  INSTALLED = 0,
+  CLONED = 1,
+  UPDATED = 2,
+  REMOVED = 3,
+  TO_INSTALL = 4,
+}
+
+--- List of packages to build
+--- @type vim._pack.Package[]
+local BuildQueue = {}
+
+--- Table of pgks loaded from the lockfile
+--- @type vim._pack.Package[]
+local Lock = {}
+
+--- Table of pkgs loaded from the user configuration
+--- @type vim._pack.Package[]
+local Packages = {}
+
+--- @type table<string, string>
+local Path = {
+  lock = vim.fs.joinpath(vim.fn.stdpath('state'), 'pack-lock.json'),
+  log = vim.fs.joinpath(vim.fn.stdpath('log'), 'pack.log'),
+  packs = vim.fs.joinpath(vim.fn.stdpath('data'), 'site', 'pack', 'packs', 'opt'),
+}
+
+--- @class vim._pack.Opts
+---
+--- Format string used to transform the package name into a git url.
+--- (default: `https://github.com/%s.git`)
+--- @field url_format string
+---
+--- Flags passed to `git clone` during installation of a plugin (see `:Man git-clone(1)` for more)
+--- (default: `{ "--depth=1", "--recurse-submodules", "--shallow-submodules", "--no-single-branch" }`)
+--- @field clone_args string[]
+---
+--- Flags passed to `git pull` during update of a plugin (see `:Man git-pull(1)` for more)
+--- (default: `{ "--tags", "--force", "--recurse-submodules", "--update-shallow" }`)
+--- @field pull_args string[]
+local Config = {
+  -- Using '--tags --force' means conflicting tags will be synced with remote
+  clone_args = { '--depth=1', '--recurse-submodules', '--shallow-submodules', '--no-single-branch' },
+  pull_args = { '--tags', '--force', '--recurse-submodules', '--update-shallow' },
+  url_format = 'https://github.com/%s.git',
+}
+
+--- @enum Messages
+local Messages = {
+  install = { ok = 'Installed', err = 'Failed to install' },
+  update = { ok = 'Updated', err = 'Failed to update', nop = '(up-to-date)' },
+  remove = { ok = 'Removed', err = 'Failed to remove' },
+  build = { ok = 'Built', err = 'Failed to build' },
+}
+
+--- @enum Filter
+local Filter = {
+  installed = function(p)
+    return p.status ~= M.status.REMOVED and p.status ~= M.status.TO_INSTALL
+  end,
+  not_removed = function(p)
+    return p.status ~= M.status.REMOVED
+  end,
+  removed = function(p)
+    return p.status == M.status.REMOVED
+  end,
+  to_install = function(p)
+    return p.status == M.status.TO_INSTALL
+  end,
+  to_update = function(p)
+    return p.status ~= M.status.REMOVED and p.status ~= M.status.TO_INSTALL and not p.pin
+  end,
+}
+
+--- @param path string
+--- @param flags string|integer
+--- @param data string
+local function file_write(path, flags, data)
+  local err_msg = "Failed to %s '" .. path .. "'"
+  local file = assert(uv.fs_open(path, flags, 0x1A4), err_msg:format('open'))
+  assert(uv.fs_write(file, data), err_msg:format('write'))
```

Comment:
```suggestion
  assert(uv.fs_write(file, data) == #data, err_msg:format('write'))
```

Unlikely (especially when writing to disk), but `fs_write` is allowed to succeed with a partial write.

Looping until everything's written or using something higher-level like a `uv_pipe_t` (though non-blocking) may be more robust, but this is probably fine.